File: android_task_queue/app/src/main/java/com/zq/taskqueue/TaskQueue.java
package com.zq.taskqueue;

import java.util.Collection;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.Map;

public class TaskQueue<T extends Task> {

    public static final long DEFAULT_MAX_TASK_COUNT = -1L;
    private final Map<String, T> taskKeyedById;
    private long maxTaskCount;
    private boolean enabled;

    public TaskQueue() {
        this.enabled = true;
        this.maxTaskCount = DEFAULT_MAX_TASK_COUNT;
        this.taskKeyedById = new LinkedHashMap<>();
    }

    public boolean contains(String id) {
        return taskKeyedById.get(id) != null;
    }

    public synchronized boolean add(T task, String id) {
        if (contains(id)) return false;
        task.set(id, () -> {
            removeTask(id);
        });
        taskKeyedById.put(id, task);
        return true;
    }

    public T get(String id) {
        return taskKeyedById.get(id);
    }

    public void performBatch(Runnable runnable) {
        if (runnable != null) {
            enabled = false;
            runnable.run();
            enabled = true;
            execute();
        }
    }

    public synchronized void startTask(String id) {
        Task task = get(id);
        if (task != null) {
            task.setEnabled(true);
            execute();
        }
    }

    public synchronized void pauseTask(String id) {
        Task task = get(id);
        if (task != null) {
            task.setEnabled(false);
            execute();
        }
    }

    public synchronized void removeTask(String id) {
        Task task = get(id);
        if (task != null) {
            task.onDispose();
            taskKeyedById.remove(id);
            execute();
        }
    }

    public void startAllTasks() {
        performBatch(() -> {
            for (T task : getTasks()) {
                task.setEnabled(true);
            }
        });
    }

    public void pauseAllTasks() {
        performBatch(() -> {
            for (T task : getTasks()) {
                task.setEnabled(false);
            }
        });
    }

    public void removeAllTasks() {
        performBatch(() -> {
            Iterator<Map.Entry<String, T>> iterator = taskKeyedById.entrySet().iterator();
            while (iterator.hasNext()) {
                Map.Entry<String, T> entry = iterator.next();
                T task = entry.getValue();
                task.onDispose();
                iterator.remove();
            }
        });
    }

    public void setMaxTaskCount(long value) {
        if (maxTaskCount != value) {
            maxTaskCount = Math.max(DEFAULT_MAX_TASK_COUNT, value);
            execute();
        }
    }

    public long getMaxTaskCount() {
        return maxTaskCount;
    }

    public Collection<T> getTasks() {
        return taskKeyedById.values();
    }

    private void execute() {
        if (!enabled) return;
        long activeCount = activeCount();
        for (T task : getTasks()) {
            if (!task.isEnabled() && task.getTaskState() == TaskState.RUNNING) {
                task.performOnPause();
            }
            if (maxTaskCount != DEFAULT_MAX_TASK_COUNT) {
                if (maxTaskCount < activeCount && task.getTaskState() == TaskState.RUNNING) {
                    task.performOnPause();
                }
            }
        }
        boolean belowMaxTask = belowMaxTask();
        for (T task : getTasks()) {
            if (belowMaxTask && task.isEnabled() && task.getTaskState() != TaskState.RUNNING) {
                task.performOnStart();
                belowMaxTask = belowMaxTask();
            }
        }
    }

    private boolean belowMaxTask() {
        if (maxTaskCount == DEFAULT_MAX_TASK_COUNT) {
            return true;
        }
        return activeCount() < maxTaskCount;
    }

    private long activeCount() {
        long count = 0;
        for (T task : getTasks()) {
            if (task.getTaskState() == TaskState.RUNNING) {
                count++;
            }
        }
        return count;
    }
}

File: ios_task_queue/ios_task_queue/Sources/TaskQueue.m
//
//  TaskQueue.m
//  TaskQueue <https://github.com/QiaokeZ/TaskQueue>
//
//  Created by zhouqiao on 2022/8/4.
//  Copyright © 2022 zhouqiao. All rights reserved.
//
//  This source code is licensed under the MIT-style license found in the
//  LICENSE file in the root directory of this source tree.
//

#import "Task.h"
#import "TaskQueue.h"
#import "Task+Private.h"

@interface TaskQueue()
@property(nonatomic) NSMutableArray<Task *> *elements;
@property(nonatomic, getter=isEnabled) BOOL enabled;
@property(nonatomic) NSMutableDictionary<NSString *, Task *> *taskKeyedByIdentifier;
@end

@implementation TaskQueue

+ (instancetype)queue {
    return [[self alloc] init];
}

- (instancetype)init {
    if (self = [super init]) {
        self.enabled = YES;
        self.elements = [NSMutableArray array];
        self.taskKeyedByIdentifier = [NSMutableDictionary dictionary];
        self.maxTaskCount = TaskQueueDefaultMaxTaskCount;
    }
    return self;
}

- (void)setMaxTaskCount:(NSInteger)maxTaskCount {
    if (_maxTaskCount != maxTaskCount) {
        _maxTaskCount = MAX(TaskQueueDefaultMaxTaskCount, maxTaskCount);
        [self execute];
    }
}

- (NSArray *)tasks {
    return self.elements.copy;
}

- (BOOL)containsIdentifier:(NSString *)identifier {
    return [self taskForIdentifier:identifier] != nil;
}

- (BOOL)addTask:(Task *)task forIdentifier:(NSString *)identifier {
    @synchronized (self) {
        if([self containsIdentifier:identifier]) return NO;
        [task _setIdentifier:identifier finishHandler:^{
            [self removeTask:identifier];
        }];
        self.taskKeyedByIdentifier[identifier] = task;
        [self.elements addObject:task];
        return YES;
    }
}

- (void)performBatchHandler:(void(^)(void))handler {
    if (handler) {
        self.enabled = NO;
        handler();
        self.enabled = YES;
        [self execute];
    }
}

- (void)startTask:(NSString *)identifier {
    @synchronized (self) {
        Task *task = [self taskForIdentifier:identifier];
        if (task) {
            task._enabled = YES;
            [self execute];
        }
    }
}

- (void)pauseTask:(NSString *)identifier {
    @synchronized (self) {
        Task *task = [self taskForIdentifier:identifier];
        if (task) {
            task._enabled = NO;
            [self execute];
        }
    }
}

- (void)removeTask:(NSString *)identifier {
    @synchronized (self) {
        Task *task = [self taskForIdentifier:identifier];
        if (task) {
            [task _dispose];
            [self.taskKeyedByIdentifier removeObjectForKey:task.identifier];
            [self.elements removeObject:task];
            [self execute];
        }
    }
}

- (void)startAllTasks {
    [self performBatchHandler:^{
        for (Task *task in self.tasks) {
            task._enabled = YES;
        }
    }];
}

- (void)pauseAllTasks {
    [self performBatchHandler:^{
        for (Task *task in self.tasks) {
            task._enabled = NO;
        }
    }];
}

- (void)removeAllTasks {
    [self performBatchHandler:^{
        [self.taskKeyedByIdentifier removeAllObjects];
        NSEnumerator *enumerator = [self.elements objectEnumerator];
        Task *obj = nil;
        while (obj = [enumerator nextObject]) {
            [obj dispose];
            [self.elements removeObject:obj];
        }
    }];
}

- (Task *)taskForIdentifier:(NSString *)identifier {
    return self.taskKeyedByIdentifier[identifier];
}

- (void)execute {
    if (!self.isEnabled) return;
    NSInteger activeCount = [self activeCount];
    for (Task *task in self.tasks) {
        if (!task.isEnabled && task.state == TaskStateRunning) {
            [task _pause];
        }
        if (self.maxTaskCount != TaskQueueDefaultMaxTaskCount) {
            if (self.maxTaskCount < activeCount && task.state == TaskStateRunning) {
                [task _pause];
            }
        }
    }
    BOOL belowMaxTask = [self belowMaxTask];
    for (Task *task in self.tasks) {
        if (belowMaxTask && task.isEnabled && task.state != TaskStateRunning) {
            [task _start];
            belowMaxTask = [self belowMaxTask];
        }
    }
}

- (BOOL)belowMaxTask {
    if (self.maxTaskCount == TaskQueueDefaultMaxTaskCount) {
        return YES;
    }
    return self.activeCount < self.maxTaskCount;
}

- (NSInteger)activeCount {
    NSInteger count = 0;
    for (NSString *key in self.taskKeyedByIdentifier) {
        Task *task = self.taskKeyedByIdentifier[key];
        if (task.state == TaskStateRunning) {
            count++;
        }
    }
    return count;
}

@end
